src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -163,6 +185,51 @@ void snc_quota_manager::record_response_tp(
     _shard_quota.out.use(request_size, now);
 }
 
+std::chrono::milliseconds
+snc_quota_manager::get_quota_balancer_node_period() const {
+    const auto v = _kafka_quota_balancer_node_period();
+    // zero period in config means do not run balancer
+    if (v == std::chrono::milliseconds::zero()) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(
+          ss::steady_clock_type::duration::max() / 2);
+    }
+    return v;
+}
+
+void snc_quota_manager::maybe_arm_balancer_timer() {
+    static constexpr auto min_sleep_time = 2ms; // TBD config?
+
+    if (_balancer_timer.armed()) {
+        return;
+    }
+    ss::steady_clock_type::time_point arm_until
+      = _balancer_timer_last_ran + get_quota_balancer_node_period();
+    const auto now = ss::steady_clock_type::now();
+    const ss::steady_clock_type::time_point closest_arm_until
+      = now + min_sleep_time;
+    if (arm_until < closest_arm_until) {
+        // occurs when balancer takes to run longer than the balancer period,
+        // and on the first iteration
+        if (
+          _balancer_timer_last_ran.time_since_epoch()
+          != ss::steady_clock_type::duration::zero()) {
+            vlog(
+              klog.warn,
+              "qb - Quota balancer is invoked too often ({}), "
+              "enforcing minimum sleep time",
+              arm_until - now);
+        }
+        arm_until = closest_arm_until;
+    }
+    _balancer_timer.arm(arm_until);
+}
+
+ss::future<> snc_quota_manager::quota_balancer_step() {
+    const auto sem_units = co_await ss::get_units(_balancer_semaphore, 1);
+    _balancer_gate.check();
```

Comment:
The `.then` continuation should be `.finally` and then the issue will manifest on each application stop. This `check()` here is exclusively to stop the balancer fiber faster in case it is unblocked from semaphore/mutex on shutdown.